modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicStore.java

```diff
@@ -0,0 +1,484 @@
+     * @param repository the repository to use for lookup of the delegate store.
+     * @param preferredDataStoreSPI the preferred SPI to use for the delegate store.
+     * @param connectionParameterKey the key to use for the connection parameter in the delegate
+     *     store.
+     */
+    public VectorMosaicStore(
+            String delegateStoreName,
+            Repository repository,
+            String preferredDataStoreSPI,
+            String connectionParameterKey) {
+        this.repository = repository;
+
+        this.preferredSPI = preferredDataStoreSPI;
+
+        if (connectionParameterKey != null) {
+            this.connectionParameterKey = connectionParameterKey;
+        }
+        Name name = buildName(delegateStoreName);
+        delegateDataStore = repository.dataStore(name);
+        try {
+            delegateTypes =
+                    Arrays.stream(delegateDataStore.getTypeNames())
+                            .filter(
+                                    t -> {
+                                        try {
+                                            return validateSchema(
+                                                    delegateStoreName,
+                                                    t,
+                                                    delegateDataStore.getSchema(t));
+                                        } catch (IOException e) {
+                                            LOGGER.log(
+                                                    Level.WARNING,
+                                                    "Failed to validate schema for type " + t,
+                                                    e);
+                                            return false;
+                                        }
+                                    })
+                            .collect(Collectors.toList());
+
+            for (String typeName : delegateTypes) {
+                addTypeName(typeName);
+            }
+
+        } catch (IOException e) {
+            LOGGER.log(
+                    Level.WARNING,
+                    "Failed to get type names for vector mosaic delegate data store "
+                            + delegateStoreName,
+                    e);
+        }
+    }
+
+    /**
+     * Adds a type name to the vector mosaic store.
+     *
+     * @param delegateTypeName the type name to add.
+     */
+    @SuppressWarnings(
+            "PMD.UseTryWithResources") // PMD complains about FeatureIterator, but it is not ready
+    // to be instantiated yet
+    private void addTypeName(String delegateTypeName) {
+        SimpleFeatureIterator iterator = null;
+        try {
+            SimpleFeatureSource featureSource =
+                    delegateDataStore.getFeatureSource(delegateTypeName);
+            SimpleFeatureCollection features = featureSource.getFeatures();
+            iterator = features.features();
+            while (iterator.hasNext()) {
```

Comment:
Oh no, keeping the granules in memory? This is a major issue:
- Invites OOMs, as the granules might be many
- Makes the system nonresponsive to granules being removed or added from the delegate store

Each time the list of granules is needed, it should be queried, taking into account the filter that's coming from the client (so pre-filter based on bbox and other properties contained in the index table, and then apply a post filter once the features in the target store are loaded).

Reply:
Fixed